File: java/Easy/_125_ValidPalindrome.java
package Easy;

import org.testng.annotations.Test;

public class _125_ValidPalindrome {
    public boolean isPalindrome(String s) {
        s = s.toLowerCase();
        char[] ar = s.toCharArray();
        StringBuilder sb = new StringBuilder();
        for(char c : ar){
            if(('a'<=c && c<='z') ||('0'<=c && c<='9') ) sb.append(c);
        }

        for(int i = 0; i<sb.length()/2; i++ ){
            if(sb.charAt(i) != sb.charAt(sb.length()-1-i)) return false;
        }

        return true;
    }

    @Test
    public void test(){
        System.out.println(isPalindrome("A man, a palan, a canal: Panama"));
    }
}

File: Python3/_125_ValidPalindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        s = s.lower()
        l = 0
        r = len(s)-1
        while l<=r:
            if ('a'<=s[l]<='z') == False and ('0'<=s[l]<='9') == False:
                l+=1
                continue
            if ('a'<=s[r]<='z') == False and ('0'<=s[r]<='9') == False:
                r-=1
                continue
            if s[l] != s[r]:
                return False
            else:
                l+=1
                r-=1
        return True
    
test = Solution()
print(test.isPalindrome("0P"))
print(test.isPalindrome("A man, a plan, a canal: Panama"))
print(test.isPalindrome("Arace a car"))